hathor/transaction/block.py

```diff
@@ -431,5 +368,6 @@ def iter_transactions_in_this_block(self) -> Iterator[BaseTransaction]:
             yield tx
 
     @override
-    def init_static_metadata_from_storage(self, storage: 'TransactionStorage') -> None:
-        raise NotImplementedError('this will be implemented')
+    def init_static_metadata_from_storage(self, settings: HathorSettings, storage: 'TransactionStorage') -> None:
```

Comment:
This is called when vertices are created/received, which can be a few places:

- When retrieving a vertex from RocksDB.
- In a few endpoints that create vertices, such as `decode_tx`.
- In tests.
- And, most importantly, in the verification service, just before performing all verifications. This is what's indirectly called for new vertices by `on_new_vertex()`.

But please note that:

1. All static metadata attributes are required for verification, so they must be calculated at that point.
2. All the same attributes were already calculated in the exact same places, either by calling `get_metadata()` or by explicitly calling one of the `update_initial_metadata()` sub-methods directly. In fact, this was a source of bugs in the past, like metadata was incorrectly initialized with `min_height=0`.
3. All calculations performed for static metadata are cheap. All of them simply access a few dependencies and perform simple calculations. There are no unbounded loops and no recursion.

Therefore, I don't think this would be a problem. That being said, it could be improved in the future so that we perform each verification after each calculation. For example, calculate the `height`, then verify it. Only if it passes, calculate the `min_height`, then verify it. And so on...

Reply:
I agree that these calculations were already there on the `get_metadata()`, so it wouldn't hurt the CPU more than it already does. But we are centralizing all these calculations in this PR so we must be careful on future changes. It seems easy to naively add new metadata fields that are expensive to be calculated.

Reply:
Sure, we must be careful. Do you have any other suggestions other than the comment that was added to the docstrings? Also relevant to mention that that's the reason I wanted to add benchmarking to the CI, to prevent this kind of mistake in the future. To make it more robust we have to increase the benchmark coverage, though.

Reply:
Thanks! A quick note: The CI benchmark would prevent us from doing something slow but not an attacker from crafting transactions that might DoS a full node.

Reply:
@msbrogli Yes, but I mean if we introduce something that is too CPU intensive, the benchmark may fail and prevent us from releasing code that would be susceptible to DoS. In other words, if a change makes more CPU use than it should, it will likely slow down the sync.